Facade blockstates & Rotation! Fixes #23

HeatherComputer opened the pull request:
Changes facades to utilise BlockStates instead of Blocks, and adds support for rotation of already placed facades.
All functionality for placing new facades currently remains the same.


As part of this, I require credit - both ingame and on CF, with the latter being fine either in the authors list (preferable) or just in the description.
I'm not expecting a revenue cut (though I certainly wouldn't complain.)

HeatherComputer committed:
yes i'm adding bin to this. it's just another buildcache

HeatherComputer committed:
Chance the chunkMap to store BlockStates instead of Blocks.
This... will cause a lot of cascading errors I'll need to fix.

HeatherComputer committed:
Codecs for BlockStates using NBT as a helper

HeatherComputer committed:
ChunkFacadeMap now uses the BlockState codec

HeatherComputer committed:
Shouldn't have removed the existing codecs... despite being unused now, they'll be important for migration?

HeatherComputer committed:
Currently migration is NOT supported! SavedData now uses BlockStates instead of Blocks.

HeatherComputer committed:
ClientFacadeManager and FacadeUtils both work with BlockStates now

HeatherComputer committed:
Networking now works with BlockStates

HeatherComputer committed:
Facade item uses the default blockstate when placing a facade.

HeatherComputer committed:
ClientGameEvents now also uses BlockStates

HeatherComputer committed:
GameEvents also works with BlockStates now

HeatherComputer committed:
import cleanup

HeatherComputer committed:
BlockMixin now handles BlockStates

HeatherComputer committed:
BlockStateBaseMixin now handles states

HeatherComputer committed:
aaand finally, the BlockStateMixin handles the changed format

HeatherComputer committed:
Import cleanup and add a TODO

HeatherComputer commented:
As of this comment - 

- Facades use BlockStates. All previous functionality works, with no user-facing changes.
- Migration is *not* yet supported. This will be changed, to avoid worlds breaking.
- Rotation is *not* yet supported. Again, this will be changed - from my understanding the preferred method is to use any wrench on an already placed facade, yes?

I do have one question here - shift-rightclick is for removing facades, which makes sense. If we hog rightclick, this would prevent a user from having whatever normal rightclick functionality the underlying pipe may have when holding a wrench - is this fine?

(As a note for the title - the issue won't be able to link itself because the PR doesn't target the default branch.)

Leclowndu93150 commented:
>  this would prevent a user from having whatever normal rightclick functionality the underlying pipe may have when holding a wrench - is this fine?

yes it is, because you have to remove the facade if you  want to change the pipe or something

HeatherComputer committed:
Update ingame authors

HeatherComputer committed:
Migration now works!

HeatherComputer committed:
Rudimentary facade rotation. More will be required for stairs etc

HeatherComputer committed:
Stop facades from double-rotating. This also lets the wrenches work in the offhand!

HeatherComputer committed:
Support rotation for anything that implements HorizontalDirectionalBlock.FACING or DirectionalBlock.FACING

Some more explicit support will be needed for the likes of stairs and slabs going upside down.

HeatherComputer commented:
As of this comment - 

- Facades use BlockStates. All previous functionality works, with no user-facing changes.
- Migration is now supported. I'm considering reworking this to be a little tidier - but it works.
- Rotation is *partially* supported. Anything that just utilises `HorizontalDirectionalBlock#FACING` or `DirectionalBlock.FACING` with no additional rotational properties works. Things like stairs may need additional adjustment.

Thepigcat76 commented on src/main/java/com/portingdeadmods/cable_facades/content/items/FacadeItem.java:
If possible, change block1.defautltBlockstate() to block1.getStateForPlacement(...)

HeatherComputer replied:
Possible, but should it be part of this PR?

Thepigcat76 replied:
I mean its just a minor convenience change, unless there is some huge impact that i cant think of rn you can include it here. Me is sleepy tho, so i havent thought about it in depth lol

HeatherComputer replied:
Upon further review, this would have the effect of the facade by default facing towards the player with rotatable blocks.
Therefore I do believe it belongs in the PR.

HeatherComputer replied:
Done

Thepigcat76 commented on src/main/java/com/portingdeadmods/cable_facades/events/GameEvents.java:
nitpick: use the modern switch statement/expression in the rotate method

HeatherComputer replied:
You make a good point. I'm too used to it just not being an option haha

I'll adjust that tomorrow, along with improving rotation support for stairs etc.

Thepigcat76 commented:
Looking pretty good! Added some smaller comments but overall, your code seems really clean. Thank you. Will of course add you to curseforge and give you a revenue split :+1:

HeatherComputer committed:
Facades will inherit a default rotation as if they were the mimicked block (so for most things, facing towards the player)

HeatherComputer committed:
Use modern switch statement in rotate method

HeatherComputer committed:
stair support

HeatherComputer committed:
shlabbed properly this time

HeatherComputer commented:
As of this comment - 

- Facades use BlockStates. All previous functionality works, with no user-facing changes.
- Migration is now supported. I'm considering reworking this to be a little tidier - but it works.
- Rotation is fully supported - stairs will cycle all directions facing up / down (no corners though) and slabs will cycle between top, bottom and full.

HeatherComputer commented:
I've marked this as ready to review - please let me know if you have any further thoughts, particularly in regards to migration.

Leclowndu93150 commented:
looks good to me

HeatherComputer commented:
Awesome. In that case I'm ready for it to be merged - for the CF side of things, I believe you need my username yeah? If so, it's `heatherwhite_`

Thepigcat76 commented on src/main/java/com/portingdeadmods/cable_facades/mixins/BlockStateBaseMixin.java:
Could you please rename camo to facade? I know you just adjusted our code, but we are trying to deprecate the word camo :3

Thepigcat76 commented on src/main/java/com/portingdeadmods/cable_facades/mixins/BlockStateMixin.java:
camo -> facade here as well pls

Thepigcat76 commented:
looking good, added some minor nitpicks

HeatherComputer committed:
Replace usages of `camo` with `facade`

HeatherComputer commented:
`camo` usages are now `facade` usages
